### PR TITLE
Money Pattern Implementation - Issue #1305

### PR DESCRIPTION
**Pull request by ImLynxie**
Money Pattern Implementation

- I've implemented the Money pattern, and provided a demonstration of how it can be used.
- This implementation is to showcase how to work with multiple types of currencies.

Pull request description

- There are certain things that I would have done differently within the demonstration aspect of the Money Pattern Implementation, i.e Error Handling, adding more checks to get rid of certain bugs, etc. but at that point I'd probably end up writing a whole banking system. 😂

### CODE REVIEW

**Comment by ToxicDreamz** [comment]
SonarCloud found 2 bugs, which I'll push a fix for.

**Review by ohbus — CHANGES_REQUESTED** [review]
Along with the bugs reported you also need to mitigate any code smell that it has and more importantly increase code coverage by providing test cases.

File: money/src/main/java/com/iluwatar/money/Account.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  public void withdrawFromBalance(Balance balance, Money moneyToWithdraw)
+          throws InsufficientFundsException {
+
+
+    if (balance.getMoney().getAmount() > moneyToWithdraw.getAmount()) {
+
+      try {
+        balance.getMoney().subtractMoneyBy(moneyToWithdraw);
+      } catch (SubtractionCannotOccurException e) {
+        System.err.println("Insufficient funds in balance.");
```
Comment (iluwatar) [inline money/src/main/java/com/iluwatar/money/Account.java:62]: Please use a logger for output. See Lombok `@Slf4j` annotation.

Comment (iluwatar) [reply money/src/main/java/com/iluwatar/money/Account.java:62]: Check this throughout the example.

File: money/src/main/java/com/iluwatar/money/Balance.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    return money;
+  }
+
+  public Account getAccount() {
+    return account;
+  }
+
+  public boolean isPrimaryBalance() {
+    return isPrimaryBalance;
+  }
```
Comment (iluwatar) [inline money/src/main/java/com/iluwatar/money/Balance.java:48]: To replace boilerplate such as these getters you're able to use Lombok annotations.

File: money/src/main/java/com/iluwatar/money/exception/BalanceCouldNotBeCreatedException.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  public BalanceCouldNotBeCreatedException(Throwable cause) {
+    super(cause);
+  }
+
+  public BalanceCouldNotBeCreatedException(String message, Throwable cause,
+                                           boolean enableSuppression,
+                                           boolean writableStackTrace) {
+    super(message, cause, enableSuppression, writableStackTrace);
+  }
+}
```
Comment (iluwatar) [inline money/src/main/java/com/iluwatar/money/exception/BalanceCouldNotBeCreatedException.java:25]: I'm not sure you need all these constructors. Maybe define only the ones you use?

**Review by iluwatar — CHANGES_REQUESTED** [review]

**Comment by ToxicDreamz** [comment]
I'll get these issues fixed as soon as possible, I just haven't had any extra time on my hands.

**Comment by ohbus** [comment]
> I'll get these issues fixed as soon as possible, I just haven't had any extra time on my hands.

No worries. Take your time and just let us know when you'll be able to complete it.

**Comment by iluwatar** [comment]
Closed due to inactivity

**Pull request closed without merging**